[For Review] Pt. 2 - Introduce TopologyMetadata to wrap InternalTopologyBuilders of named topologies

ableegoldman opened the pull request:
Same as apache#10683 but targeted against apache#10609 to facilitate review and filter out everything except for the changes in this PR alone

ableegoldman committed:
Implement TopologyMetadata

mdedetrich committed:
KAFKA-12819: Add assert messages to MirrorMaker tests plus other quality of life improvements (#10762)

Reviewers: Ryanne Dolan <ryannedolan@gmail.com>, Konstantine Karantasis <k.karantasis@gmail.com>

jlprat committed:
KAFKA-12796: Removal of deprecated classes under streams-scala (#10710)

Removes previously deprecated methods in older KIPs

Reviewers: Bruno Cadonna <cadonna@apache.org>

ijuma committed:
MINOR: Adjust parameter ordering of `waitForCondition` and `retryOnExceptionWithTimeout` (#10759)

New parameters in overloaded methods should appear later apart from
lambdas that should always be last.

kowshik committed:
MINOR: Improve Log layer segment iteration logic and few other areas (#10684)

In Log.collectAbortedTransactions() I've restored a previously used logic, such that it would handle the case where the starting segment could be null. This was the case previously, but the PR #10401 accidentally changed the behavior causing the code to assume that the starting segment won't be null.

In Log.rebuildProducerState() I've removed usage of the allSegments local variable. The logic looks a bit simpler after I removed it.

I've introduced a new LogSegments.higherSegments() API. This is now used to make the logic a bit more readable in Log. collectAbortedTransactions() and Log.deletableSegments() APIs.

I've removed the unnecessary use of java.lang.Long in LogSegments class' segments map definition.

I've converted a few LogSegments API from public to private, as they need not be public.

Reviewers: Ismael Juma <ismael@juma.me.uk>, Cong Ding <cong@ccding.com>, Jun Rao <junrao@gmail.com>

wcarlson5 commented on streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java:
```diff
@@ -959,23 +948,6 @@ private static Metrics getMetrics(final StreamsConfig config, final Time time, f
         return new Metrics(metricConfig, reporters, time, metricsContext);
     }
 
-    private int getNumStreamThreads(final boolean hasGlobalTopology) {
```
very glad to get rid of this!

wcarlson5 commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopologyBuilder.java:
```diff
@@ -345,8 +343,13 @@ private SinkNodeFactory(final String name,
         }
     }
 
+    public void setTopologyName(final String namedTopology) {
```
Do we really want to be able to rename a topology? maybe this can only have effect if the name is not set, or throw an exception if otherwise?

ableegoldman replied:
Oh man, this was a whole thing. No we definitely don't want to be able to rename them, but due to various java shenanigans it's pretty much impossible to get this name passed in to the constructor (it comes down to being unable to set `this.topologyName = topologyName` before calling `super();` in the `NamedTopologyStreamsBuilder` constructor, since its superclass directly instantiates the InternalTopologyBuilder. However, I'll add a check and make sure it's not possible to accidentally re-set the name once it's been set

wcarlson5 replied:
I think the check should be sufficient

wcarlson5 commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopologyBuilder.java:
```diff
@@ -1039,13 +1046,22 @@ private void buildProcessorNode(final Map<String, ProcessorNode<?, ?, ?, ?>> pro
                     // remember the changelog topic if this state store is change-logging enabled
                     if (stateStoreFactory.loggingEnabled() && !storeToChangelogTopic.containsKey(stateStoreName)) {
-                        final String changelogTopic = ProcessorStateManager.storeChangelogTopic(applicationId, stateStoreName);
+                        final String changelogTopic = ProcessorStateManager.storeChangelogTopic(applicationId, stateStoreName, namedTopology);
```
supper nit: I think this line might be too long

wcarlson5 commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopologyBuilder.java:
```diff
@@ -1213,39 +1237,26 @@ private void setRegexMatchedTopicToStateStore() {
-        if (builder.length() > 0) {
-            builder.setLength(builder.length() - 1);
-            return Pattern.compile(builder.toString());
+    public OffsetResetStrategy offsetResetStrategy(final String topic) {
```
Are these for only internal topics? if so can the name reflect that?

ableegoldman replied:
No, its for all source topics. But I can rename to `getOffsetResetStrategyForSourceTopic` (or any suggestion you have)

ableegoldman replied:
Mm actually let's leave the method renaming for another time, there's enough renaming/substituting in this PR as it is 😬

wcarlson5 replied:
as long as we have a reason

wcarlson5 commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopologyBuilder.java:
```diff
@@ -1923,7 +1970,12 @@ public void addGlobalStore(final TopologyDescription.GlobalStore globalStore) {
+            if (namedTopology == null) {
+                sb.append("Topologies:\n ");
+            } else {
+                sb.append(namedTopology).append(":\n ");
```
Do we want a prefix for readability here or later?

ableegoldman replied:
I'm not sure I'm understanding the question here, can you clarify?

wcarlson5 replied:
something like `Topology: {name}`

ableegoldman replied:
Ah, I see. Yeah that sounds like a good idea, I'll put that in

ijuma committed:
KAFKA-12856: Upgrade Jackson to 2.12.3 (#10778)

2.10.x is no longer supported, so we should move to 2.12 for the 3.0
release.

ScalaObjectMapper has been deprecated and it looks like we don't
actually need it, so remove its usage.

Reviewers: David Jacot <djacot@confluent.io>

ijuma committed:
MINOR: Don't ignore deletion of partition metadata file and log topic id clean-ups (#10761)

Log if deletion fails and don't expose log topic id for mutability outside of `assignTopicId()`.

Also remove an unnecessary parameter in `PartitionTest`.

Reviewers: Chia-Ping Tsai <chia7712@gmail.com>, Justine Olshan <jolshan@confluent.io>

lbradstreet committed:
KAFKA-12791: ConcurrentModificationException in AbstractConfig use by KafkaProducer (#10704)

Recently we have noticed multiple instances where KafkaProducers have failed to constructor due to the following exception:

```
org.apache.kafka.common.KafkaException: Failed to construct kafka producer at 
org.apache.kafka.clients.producer.KafkaProducer.<init>(KafkaProducer.java:440) at 
org.apache.kafka.clients.producer.KafkaProducer.<init>(KafkaProducer.java:291) at 
org.apache.kafka.clients.producer.KafkaProducer.<init>(KafkaProducer.java:318) 
java.base/java.lang.Thread.run(Thread.java:832) Caused by: java.util.ConcurrentModificationException at 
java.base/java.util.HashMap$HashIterator.nextNode(HashMap.java:1584) at 
java.base/java.util.HashMap$KeyIterator.next(HashMap.java:1607) at 
java.base/java.util.AbstractSet.removeAll(AbstractSet.java:171) at 
org.apache.kafka.common.config.AbstractConfig.unused(AbstractConfig.java:221) at 
org.apache.kafka.common.config.AbstractConfig.logUnused(AbstractConfig.java:379) at 
org.apache.kafka.clients.producer.KafkaProducer.<init>(KafkaProducer.java:433) ... 9 more 
exception.class:org.apache.kafka.common.KafkaException exception.message:Failed to construct kafka producer
```

This is due to the fact that `used` below is a synchronized set. `used` is being modified while removeAll is being called. This is due to the use of RecordingMap in the Sender thread (see below). Switching to a ConcurrentHashSet avoids this issue as it support concurrent iteration.

```
	at org.apache.kafka.clients.producer.ProducerConfig.ignore(ProducerConfig.java:569)
	at org.apache.kafka.common.config.AbstractConfig$RecordingMap.get(AbstractConfig.java:638)
	at org.apache.kafka.common.network.ChannelBuilders.createPrincipalBuilder(ChannelBuilders.java:242)
	at org.apache.kafka.common.network.PlaintextChannelBuilder$PlaintextAuthenticator.<init>(PlaintextChannelBuilder.java:96)
	at org.apache.kafka.common.network.PlaintextChannelBuilder$PlaintextAuthenticator.<init>(PlaintextChannelBuilder.java:89)
	at org.apache.kafka.common.network.PlaintextChannelBuilder.lambda$buildChannel$0(PlaintextChannelBuilder.java:66)
	at org.apache.kafka.common.network.KafkaChannel.<init>(KafkaChannel.java:174)
	at org.apache.kafka.common.network.KafkaChannel.<init>(KafkaChannel.java:164)
	at org.apache.kafka.common.network.PlaintextChannelBuilder.buildChannel(PlaintextChannelBuilder.java:79)
	at org.apache.kafka.common.network.PlaintextChannelBuilder.buildChannel(PlaintextChannelBuilder.java:67)
	at org.apache.kafka.common.network.Selector.buildAndAttachKafkaChannel(Selector.java:356)
	at org.apache.kafka.common.network.Selector.registerChannel(Selector.java:347)
	at org.apache.kafka.common.network.Selector.connect(Selector.java:274)
	at org.apache.kafka.clients.NetworkClient.initiateConnect(NetworkClient.java:1097)
	at org.apache.kafka.clients.NetworkClient.access$700(NetworkClient.java:87)
	at org.apache.kafka.clients.NetworkClient$DefaultMetadataUpdater.maybeUpdate(NetworkClient.java:1276)
	at org.apache.kafka.clients.NetworkClient$DefaultMetadataUpdater.maybeUpdate(NetworkClient.java:1164)
	at org.apache.kafka.clients.NetworkClient.poll(NetworkClient.java:637)
	at org.apache.kafka.clients.producer.internals.Sender.runOnce(Sender.java:327)
	at org.apache.kafka.clients.producer.internals.Sender.run(Sender.java:242)
```

Reviewers: Ismael Juma <ismael@juma.me.uk>

ableegoldman committed:
Fix up javadocs, some small stuff

wcarlson5 commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/TopologyMetadata.java:
```diff
@@ -16,10 +16,298 @@
+
+    public final void rewriteAndBuildTopology(final StreamsConfig config) {
+        // As we go, check each topology for overlap in the set of input topics/patterns
+        final Set<String> allInputTopics = new HashSet<>();
```
Do we need to remake this each time or can we track it for the whole object?

ableegoldman replied:
Well we should only need to call this once -- for now. As it happens I did actually move this to a class field in Pt. 3 since by then we'll be adding new topologies that need to be verified

wcarlson5 commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/namedtopology/KafkaStreamsNamedTopologyWrapper.java:
```diff
@@ -16,27 +16,83 @@
+     */
+    public KafkaStreamsNamedTopologyWrapper(final List<NamedTopology> topologies, final Properties props, final KafkaClientSupplier clientSupplier) {
+        super(
+            new TopologyMetadata(topologies.stream().collect(Collectors.toMap(
```
could/should we sort these based on names, that way we don't have to worry about assignments shifting a streams object is remade?

wcarlson5 replied:
I suppose that might be a issue for ones added one at a time, but if we kept it sorted things might be simpler

ableegoldman replied:
Yep, I totally agree -- that's why I made this into a TreeMap/SortedMap

wcarlson5 replied:
Ah, I didn't make that connection. So if order of the list passed into the ctor doesn't matter perhaps it should be a collection?

ableegoldman replied:
Fair enough

ableegoldman commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopologyBuilder.java:
```diff
@@ -1318,26 +1329,48 @@ public String decoratePseudoTopic(final String topic) {
         }
     }
 
+    private String buildSourceTopicsPatternString() {
```
Changed from returning a Pattern directly to returning the pattern string, so it's easier to combine multiple pattern strings from different named topologies if necessary (exposed via `sourceTopicsPatternString()` below)

ableegoldman commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/TopologyMetadata.java:
```diff
@@ -16,10 +16,298 @@
+//  3) move implementation of getter-like methods to this class, refactor so InternalTopologyBuilder is just the builder
+//  4) make sure all task-level logging includes NamedTopology
 
 public class TopologyMetadata {
```
This is the main work of this PR -- basically each NamedTopology retains its own InternalTopologyBuilder, and the new TopologyMetadata class wraps them all and handles consolidating the metadata, verifying constraints (eg that input topics are unique), and so on. Most of the changes here are just refactoring all the many things that directly depended on the InternalTopologyBuilder to instead go through this class, and then refactoring InternalTopologyBuilder to better fit in the new framework

ableegoldman commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/namedtopology/KafkaStreamsNamedTopologyWrapper.java:
```diff
@@ -16,27 +16,84 @@
+ *
+ * Status: basic architecture implemented but no actual upgrades are supported yet
+ *
+ * Note: some standard features of Kafka Streams are not yet supported with NamedTopologies. These include:
+ *       - global state stores
+ *       - interactive queries (IQ)
+ *       - TopologyTestDriver (TTD)
```
Just want to call this out in case you come across one of these things and wonder if it will work with Named Topologies -- it won't 🙂

wcarlson5 reviewed (commented):
I finished a first pass, I didn't have a ton to say about this. A lot of renaming

ableegoldman committed:
review feedback

ableegoldman committed:
fixing up tests, switch to StreamsConfig

chia7712 committed:
MINOR: make sure all fiedls of o.p.k.s.a.Action are NOT null (#10764)

I'm migrating Ranger's kafka plugin from deprecated Authorizer (this is already removed by 976e78e) to new API (see https://issues.apache.org/jira/browse/RANGER-3231). The kafka plugin needs to take something from field resourcePattern but it does not know whether the field is nullable (or users need to add null check). I check all usages and I don't observe any null case.

Reviewers: Ismael Juma <ismael@juma.me.uk>

jlprat committed:
KAFKA-12782: Fix Javadocs generation by upgrading JDK (#10780)

This, upgrades JDK to version 15 for the docs generation, this way we
can circumvent bug https://bugs.openjdk.java.net/browse/JDK-8215291
present in JDK11

Reviewers: Ismael Juma <ismael@juma.me.uk>

vvcephei committed:
KAFKA-8410: KTableProcessor migration groundwork (#10744)

* Lay the groundwork for migrating KTable Processors to the new PAPI.
* Migrate the KTableFilter processor to prove that the groundwork works.

This is an effort to help break up #10507 into multiple PRs.

Reviewers: Boyang Chen <boyang@apache.org>

chia7712 committed:
MINOR: upgrade pip from 20.2.2 to 21.1.1 (#10661)

The following error happens on my mac m1 when building docker image for system tests.

Collecting pynacl
  Using cached PyNaCl-1.4.0.tar.gz (3.4 MB)
  Installing build dependencies ... error
  ERROR: Command errored out with exit status 1:
   command: /usr/bin/python3 /usr/local/lib/python3.8/dist-packages/pip install --ignore-installed --no-user --prefix /tmp/pip-build-env-k867aac0/overlay --no-warn-script-location --no-binary :none: --only-binary :none: -i https://pypi.org/simple -- 'setuptools>=40.8.0' wheel 'cffi>=1.4.1; python_implementation != '"'"'PyPy'"'"''
       cwd: None
  Complete output (14 lines):
  Traceback (most recent call last):
    File "/usr/lib/python3.8/runpy.py", line 194, in _run_module_as_main
      return _run_code(code, main_globals, None,
    File "/usr/lib/python3.8/runpy.py", line 87, in _run_code
      exec(code, run_globals)
    File "/usr/local/lib/python3.8/dist-packages/pip/__main__.py", line 23, in <module>
      from pip._internal.cli.main import main as _main  # isort:skip # noqa
    File "/usr/local/lib/python3.8/dist-packages/pip/_internal/cli/main.py", line 5, in <module>
      import locale
    File "/usr/lib/python3.8/locale.py", line 16, in <module>
      import re
    File "/usr/lib/python3.8/re.py", line 145, in <module>
      class RegexFlag(enum.IntFlag):
  AttributeError: module 'enum' has no attribute 'IntFlag'
  ----------------------------------------
ERROR: Command errored out with exit status 1: /usr/bin/python3 /usr/local/lib/python3.8/dist-packages/pip install --ignore-installed --no-user --prefix /tmp/pip-build-env-k867aac0/overlay --no-warn-script-location --no-binary :none: --only-binary :none: -i https://pypi.org/simple -- 'setuptools>=40.8.0' wheel 'cffi>=1.4.1; python_implementation != '"'"'PyPy'"'"'' Check the logs for full command output.

There was a related issue: pypa/pip#9689 and it is already fixed by pypa/pip#9689 (included by pip 21.1.1). I test the pip 21.1.1 and it works well on mac m1.

Reviewers: Ismael Juma <ismael@juma.me.uk>

rohit-sachan committed:
KAFKA-12865 : Fix doc error in Admin.describeAcls API (#10790)

Reviewers: Lee Dongjin <dongjin@apache.org>, Manikumar Reddy <manikumar.reddy@gmail.com>

ijuma committed:
MINOR: Reduce allocations in requests via buffer caching (#9229)

Use a caching `BufferSupplier` per request handler thread so that
decompression buffers are cached if supported by the underlying
`CompressionType`. This achieves a similar outcome as #9220, but
with less contention.

We introduce a `RequestLocal` class to make it easier to introduce
new request scoped stateful instances (one example we discussed
previously was an `ActionQueue` that could be used to avoid
some of the complex group coordinator locking).

This is a small win for zstd (no synchronization or soft references) and
a more significant win for lz4. In particular, it reduces allocations
significantly when the number of partitions is high. The decompression
buffer size is typically 64 KB, so a produce request with 1000 partitions
results in 64 MB of allocations even if each produce batch is small (likely,
when there are so many partitions).

I did a quick producer perf local test with 5000 partitions, 1 KB record
size,
1 broker, lz4 and ~0.5 for the producer compression rate metric:

Before this change:
> 20000000 records sent, 346314.349535 records/sec (330.27 MB/sec),
148.33 ms avg latency, 2267.00 ms max latency, 115 ms 50th, 383 ms 95th, 777 ms 99th, 1514 ms 99.9th.

After this change:
> 20000000 records sent, 431956.113259 records/sec (411.95 MB/sec),
117.79 ms avg latency, 1219.00 ms max latency, 99 ms 50th, 295 ms 95th, 440 ms 99th, 662 ms 99.9th.

That's a 25% throughput improvement and p999 latency was reduced to
under half (in this test).

Default arguments will be removed in a subsequent PR.

Reviewers: Chia-Ping Tsai <chia7712@gmail.com>

chia7712 committed:
HOTFIX: fix build error (#10796)

Fix compile error in scala tests.

The compile error is:

```
[Error] /home/jenkins/jenkins-agent/workspace/Kafka_kafka-pr_PR-9229/core/src/test/scala/unit/kafka/server/metadata/BrokerMetadataListenerTest.scala:97: polymorphic expression cannot be instantiated to expected type;

[2021-05-29T02:34:50.308Z]  found   : [T]()T

[2021-05-29T02:34:50.308Z]  required: kafka.server.RequestLocal
```

This error happens only in scala 2.12

Reviewers: Bruno Cadonna <cadonna@apache.org>

jlprat committed:
MINOR: Dependency updates for Scala libraries for improved Scala 3.0 support (#10783)

Release notes:
* Scala 2.12.14: https://github.com/scala/scala/releases/tag/v2.12.14
* Scala Logging: https://github.com/lightbend/scala-logging/releases/tag/v3.9.3
* Scala Collection Compat:
  *  https://github.com/scala/scala-collection-compat/releases/tag/v2.3.1
  * https://github.com/scala/scala-collection-compat/releases/tag/v2.3.2
  * https://github.com/scala/scala-collection-compat/releases/tag/v2.4.0
  * https://github.com/scala/scala-collection-compat/releases/tag/v2.4.1
  * https://github.com/scala/scala-collection-compat/releases/tag/v2.4.2
  * https://github.com/scala/scala-collection-compat/releases/tag/v2.4.3
  * https://github.com/scala/scala-collection-compat/releases/tag/v2.4.4
* Scala Java8 Compat:
  * https://github.com/scala/scala-java8-compat/releases/tag/v1.0.0-RC1
  * https://github.com/scala/scala-java8-compat/releases/tag/v1.0.0

Reviewers: Ismael Juma <ismael@juma.me.uk>

chia7712 committed:
MINOR: cleanTest ought to remove output of unitTest task and integrat… (#10585)

The command used by our private CI is ./gradlew cleanTest xxx:test. It does not re-run test when we use unitTest and integrationTest to replace test. The root cause is that we don't offer test output (unitTest and integrationTest) to cleanTest task and so it does not delete related test output.

Reviewers: Ismael Juma <ismael@juma.me.uk>

chia7712 committed:
MINOR: replace by org.junit.jupiter.api.Tag by net.jqwik.api.Tag for raft module (#10791)

The command `./gradlew raft:integrationTest`  can't run any integration test since `org.junit.jupiter.api.Tag` does not work for jqwik engine (see https://github.com/jlink/jqwik/issues/36#issuecomment-436535760). 

Reviewers: Ismael Juma <ismael@juma.me.uk>

jlprat committed:
MINOR: fix code listings security.html (#10770)

Fix examples under security.html so they use the right bash icon (`>`
instead of `$`) and also uses the right tool for showing code listings.

Reviewers: Luke Chen <showuon@gmail.com>, Bruno Cadonna <cadonna@apache.org>

cadonna committed:
KAFKA-12519: Remove built-in Streams metrics for versions 0.10.0-2.4 (#10765)

As specified in KIP-743, this PR removes the built-in metrics
in Streams that are superseded by the refactoring proposed in KIP-444.

Reviewers: Guozhang Wang <wangguoz@gmail.com>, Luke Chen <showuon@gmail.com>

soarez committed:
KAFKA-12866: Avoid root access to Zookeeper (#10795)

The broker shouldn't assume create access to the chroot. There are
deployement scenarios where the chroot is already created is the only
znode which the broker can access.

Reviewers: Manikumar Reddy <manikumar.reddy@gmail.com>, Ron Dagostino <rdagostino@confluent.io>

showuon committed:
KAFKA-12596: remove --zookeeper option from topic command (#10457)

Also:
* Remove `ZookeeperTopicService`
* Remove `TopicCommandWithZKClientTest`
* Fix a topic create validation bug
* Adjust existing tests

Reviewers: Ismael Juma <ismael@juma.me.uk>

hachikuji committed:
KAFKA-12709; Add Admin API for `ListTransactions` (#10616)

This patch adds `Admin` support for the `listTransactions` API, which was added by [KIP-664](https://cwiki.apache.org/confluence/display/KAFKA/KIP-664%3A+Provide+tooling+to+detect+and+abort+hanging+transactions). Similar to `listConsumerGroups`, the new `listTransactions` API is intended to be sent to all brokers. 

Reviewers: David Jacot <djacot@confluent.io>

KahnCheny committed:
MINOR: remove unneccessary public keyword from ProducerInterceptor/ConsumerInterceptor interface (#10801)

Co-authored-by: “KahnCheny” <“kahn.cheny@gmail.com”>

Reviewers: Luke Chen <showuon@gmail.com>, David Jacot <djacot@confluent.io>

ijuma committed:
MINOR: Update jmh for async profiler 2.0 support (#10800)

Async profiler 2.0 outputs html5 flame graph files
and supports simultaneous collection of cpu,
allocation and lock profiles in jfr format.

Updated the readme to include an example of the
latter and verified that the Readme commands
work with async profiler 2.0.

Release notes:
* 1.28: https://mail.openjdk.java.net/pipermail/jmh-dev/2021-March/003171.html
* 1.29: https://mail.openjdk.java.net/pipermail/jmh-dev/2021-March/003218.html
* 1.30: https://mail.openjdk.java.net/pipermail/jmh-dev/2021-May/003237.html
* 1.31: https://mail.openjdk.java.net/pipermail/jmh-dev/2021-May/003286.html
* 1.32: https://mail.openjdk.java.net/pipermail/jmh-dev/2021-May/003307.html

Reviewers: Chia-Ping Tsai <chia7712@gmail.com>, David Jacot <djacot@confluent.io>, Luke Chen <showuon@gmail.com>

wenbingshen committed:
MINOR: Update kafka-topics.sh line command tool upgrade notes with removed option (#10806)

Reviewers: Luke Chen <showuon@gmail.com>, Ismael Juma <ismael@juma.me.uk>

Co-authored-by: shenwenbing <shenwenbing@qianxin.com>

kowshik committed:
KAFKA-12867: Fix ConsumeBenchWorker exit behavior for maxMessages config (#10797)

The trogdor ConsumeBenchWorker allows several consumption tasks to be run in parallel, the number is configurable using the threadsPerWorker config. If one of the consumption tasks completes executing successfully due to maxMessages being consumed, then, the consumption task prematurely notifies the doneFuture causing the entire ConsumeBenchWorker to halt. This becomes a problem when more than 1 consumption task is running in parallel, because the successful completion of 1 of the tasks shuts down the entire worker while the other tasks are still running. When the worker is shut down, it kills all the active consumption tasks, though they have not consumed maxMessages yet. This commit defers notification of the doneFuture to the CloseStatusUpdater thread, which is already responsible for tracking the status of the tasks and updating their status when all of the tasks complete.

Reviewers: Rajini Sivaram <rajinisivaram@googlemail.com>

ijuma committed:
KAFKA-12880: Remove deprecated `Count` and `SampledTotal` in 3.0 (#10808)

They were both deprecated in Apache Kafka 2.4 and it's a straightforward change
to use the non deprecated variants.

Reviewers: David Jacot <djacot@confluent.io>

showuon committed:
KAFKA-12675: improve the sticky general assignor scalability and performance (#10552)

I did code refactor/optimization, keep the same algorithm in this PR.

Originally, With this setting:
topicCount = 50;
partitionCount = 800;
consumerCount = 800;
We complete in 10 seconds, after my code refactor, the time down to 100~200 ms

With the 1 million partitions setting:
topicCount = 500;
partitionCount = 2000;
consumerCount = 2000;
No OutOfMemory will be thrown anymore. The time will take 4~5 seconds.

Reviewers: Vahid Hashemian <vahid.hashemian@gmail.com>, Guozhang Wang <wangguoz@gmail.com>

cmccabe committed:
KAFKA-12864: Move KafkaEventQueue into server-common. #10787 (#10787)

Since KafkaEventQueue is a generic data structure not specific to metadata, move it
into the server-common module.

Reviewers: Ismael Juma <ismael@juma.me.uk>, David Arthur <mumrah@gmail.com>

Commit:
MINOR: Apply try-with-resource to KafkaStreamsTest (#10668)

Reviewers: Chia-Ping Tsai <chia7712@gmail.com>, Matthias J. Sax <matthias@confluent.io>

vichu committed:
KAFKA-12749: Changelog topic config on suppressed KTable lost (#10664)

Refactored logConfig to be passed appropriately when using shutDownWhenFull or emitEarlyWhenFull. Removed the constructor that doesn't accept a logConfig parameter so you're forced to specify it explicitly, whether it's empty/unspecified or not.

Co-authored-by: Bruno Cadonna <cadonna@apache.org>

Reviewers: Walker Carlson <wcarlson@confluent.io>, Bruno Cadonna <cadonna@apache.org>

mumrah committed:
KAFKA-12620 Allocate Producer IDs in KRaft controller (#10752)

This is part 2 of KIP-730. Part 1 was in #10504.

This PR adds QuorumController support for handling AllocateProducerIDs requests
and managing the state of the latest producer ID block in the controller by committing
this state to the metadata log.

Reviewers: Colin P. McCabe <cmccabe@apache.org>

Commit:
minor stylish fixes to raft client (#10809)

Style fixes to KafkaRaftClient

Reviewers: Luke Chen <showuon@gmail.com>

showuon committed:
KAFKA-12597: Remove deprecated --zookeeper option in ReassignPartitionsCommand (#10471)

Also remove zookeeper dependent methods and tests.

Reviewers: Ismael Juma <ismael@juma.me.uk>

guozhangwang committed:
KAFKA-8897 Follow-up: Consolidate the global state stores (#10646)

1. When register state stores, add the store to globalStateStores before calling any blocking calls that may throw errors, so that upon closing we would close the stores as well.
2. Remove the other list as a local field, and call topology.globalStateStores when needed to get the list.

Reviewers: A. Sophie Blee-Goldman <sophie@confluent.io>, Bruno Cadonna <cadonna@apache.org>

kowshik committed:
MINOR: LogLoader: Add log identifier at few missing areas (#10819)

Reviewers: Jun Rao <junrao@gmail.com>

soarez committed:
KAFKA-12892: Use dedicated root in ZK ACL test (#10821)

Having the `testChrootExistsAndRootIsLocked` test in a separate `ZookeeperTestHarness` isn't enough to prevent the ACL changes to the ZK root from affecting other integration tests. So instead, let's use a dedicated znode for this test. It still works because `makeSurePersistentPathExists` uses `createRecursive`, which will recurse and act the same for the root or a given znode.

 Reviewers: Manikumar Reddy <manikumar.reddy@gmail.com>

dajac committed:
MINOR: Log member id of the leader when assignment are received (#10817)

Reviewers: Luke Chen <showuon@gmail.com>, Jason Gustafson <jason@confluent.io>

tombentley committed:
KAFKA-10614: Ensure group state (un)load is executed in the right order (#9441)

Co-authored-by: Jason Gustafson<jason@confluent.io>
Reviewers: Jason Gustafson<jason@confluent.io>, Guozhang Wang <wangguoz@gmail.com>

dengziming committed:
KAFKA-12338; Remove unused `MetadataParser` (#10793)

`MetadataParser` is a duplication of `MetadataRecordSerde` and it's not used in any code, so we can remove it. It did, however, have some useful validations which have been moved into `MetadataRecordSerde`.

Reviewers: Jason Gustafson <jason@confluent.io>

rondagostino committed:
KAFKA-12897: KRaft multi-partition placement on single broker (#10823)

#10494 introduced a bug in the KRaft controller where the controller will loop forever in StripedReplicaPlacer trying to identify the racks on which to place partition replicas if there is a single unfenced broker in the cluster and the number of requested partitions in a CREATE_TOPICS request is greater than 1.

This patch refactors out some argument sanity checks and invokes those checks in both RackList and StripedReplicaPlacer, and it adds tests for this as well as the single broker placement issue.

Reviewers: Jun Rao <junrao@gmail.com>

ableegoldman committed:
KAFKA-12648: Pt. 1 - Add NamedTopology to protocol and state directory structure (#10609)

This PR includes adding the NamedTopology to the Subscription/AssignmentInfo, and to the StateDirectory so it can place NamedTopology tasks within the hierarchical structure with task directories under the NamedTopology parent dir.

Reviewers: Walker Carlson <wcarlson@confluent.io>, Guozhang Wang <guozhang@confluent.io>

ableegoldman committed:
merge with upstream

ableegoldman commented:
Closing this since Pt. 1 is now merged, please move discussion to https://github.com/apache/kafka/pull/10683
